shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java

```diff
@@ -1,178 +1,245 @@
+          null,
+          dartEntrypointFunctionName
+      );
     }
 
-    int replyId = 0;
-    if (callback != null) {
-      replyId = mNextReplyId++;
-      mPendingReplies.put(replyId, callback);
-    }
-    if (message == null) {
-      flutterJNI.nativeDispatchEmptyPlatformMessage(nativeObjectReference, channel, replyId);
-    } else {
-      flutterJNI.nativeDispatchPlatformMessage(
-          nativeObjectReference, channel, message, message.position(), replyId);
+    public DartEntrypoint(
+        @NonNull AssetManager androidAssetManager,
+        @NonNull String pathToBundleWithDartEntrypoint,
+        @Nullable String pathToFallbackBundle,
+        @NonNull String dartEntrypointFunctionName
+    ) {
+      this.androidAssetManager = androidAssetManager;
+      this.pathToBundleWithDartEntrypoint = pathToBundleWithDartEntrypoint;
+      this.pathToFallbackBundle = pathToFallbackBundle;
+      this.dartEntrypointFunctionName = dartEntrypointFunctionName;
     }
   }
 
-  // Called by native to send us a platform message.
-  @Override
-  public void handlePlatformMessage(final String channel, byte[] message, final int replyId) {
-    // TODO(mattcarroll): how can we possibly be detached from JNI when this method is invoked from JNI?
-    assertAttachedToJni();
-    BinaryMessenger.BinaryMessageHandler handler = messageHandlers.get(channel);
-    if (handler != null) {
-      try {
-        final ByteBuffer buffer = (message == null ? null : ByteBuffer.wrap(message));
-        handler.onMessage(buffer, new BinaryMessenger.BinaryReply() {
-          private final AtomicBoolean done = new AtomicBoolean(false);
-          @Override
-          public void reply(ByteBuffer reply) {
-            if (!isAttachedToJni) {
-              Log.d(TAG,
-                  "handlePlatformMessage replying to a dart executor that is detached from JNI, channel="
-                      + channel);
-              return;
-            }
-            if (done.getAndSet(true)) {
-              throw new IllegalStateException("Reply already submitted");
-            }
-            if (reply == null) {
-              flutterJNI.nativeInvokePlatformMessageEmptyResponseCallback(nativeObjectReference, replyId);
-            } else {
-              flutterJNI.nativeInvokePlatformMessageResponseCallback(nativeObjectReference, replyId, reply, reply.position());
-            }
-          }
-        });
-      } catch (Exception ex) {
-        Log.e(TAG, "Uncaught exception in binary message listener", ex);
-        flutterJNI.nativeInvokePlatformMessageEmptyResponseCallback(nativeObjectReference, replyId);
-      }
-      return;
+  /**
+   * Configuration options that specify which Dart callback function is executed and where
+   * to find that callback and other assets required for Dart execution.
+   */
+  public static class DartCallback {
+    public final AssetManager androidAssetManager;
+    public final String pathToBundleWithDartEntrypoint;
```

Comment:
There's no requirement that the first bundle contains entry point. There could be a use case when the entry point is in the second bundle. I think a better way to name them would be pathToDefaultBundle and pathToFallbackBundle, which makes it clear in what order they are used to look up the assets...